### PR TITLE
Add pushdown of PostgreSQL LIMIT clause #2

### PR DESCRIPTION
**Pull request by darold**
The clause is translated into FETCH FIRST n ROWS ONLY in remote query sent to Oracle. There is two restrictions to the push down of the LIMIT clause, an OFFSET clause must not be present in the query and the minimum Oracle version must not be 12c.

### CODE REVIEW

**Comment by laurenz** [comment]
I'll have a look at it as soon as I can. I am quite busy these days, lots of teaching.

File: oracle_fdw.c [hunk 1]
```diff
@@ -732,6 +734,69 @@ _PG_init(void)
...
+	{
+		if (IsA(root->parse->limitCount, Const) && !((Const *) root->parse->limitCount)->constisnull)
+		{
+			int limit_at_end = false;
+
+			/*
+			 * we must be sure that there is no ORDER BY clause that
+			 * appears after the LIMIT clause otherwise we will have
+			 * wrong results.
+			 */
```
Comment (laurenz) [inline oracle_fdw.c:759]: I still don't understand this.
How can an `ORDER BY` be after a `LIMIT`?
Perhaps the code is right, but the comment should explain that.

File: oracle_fdw.c [hunk 2]
```diff
@@ -732,6 +734,69 @@ _PG_init(void)
...
+			if (root->parse->sortClause != NULL)
+			{
+				ListCell   *lc;
+				foreach (lc, root->processed_tlist)
+				{
+					Node *ptl = (Node *) lfirst(lc);
+					if (IsA(ptl, TargetEntry))
+					{
+						TargetEntry *tge = (TargetEntry *) ptl;
+						/* Aggregate function are pushed down so it is safe to continue */
```
Comment (laurenz) [inline oracle_fdw.c:769]: Aggregate functions are *not* pushed down, and indeed the last test case shows that the patch is doing the wrong thing.
If there is a `GROUP BY` clause, you cannot push down the `LIMIT`, because `LIMIT` is calculated after `GROUP BY`.

File: oracle_fdw.c [hunk 3]
```diff
@@ -771,11 +836,6 @@ oracleGetForeignRelSize(PlannerInfo *root, RelOptInfo *baserel, Oid foreigntable
 								&(fdwState->local_conds),
 								&(fdwState->remote_conds)
 							);
-
-	/* release Oracle session (will be cached) */
-	pfree(fdwState->session);
-	fdwState->session = NULL;
-
```
Comment (laurenz) [inline oracle_fdw.c:778]: Why did you remove that?

File: oracle_fdw.c [hunk 4]
```diff
@@ -890,6 +950,23 @@ oracleGetForeignPaths(PlannerInfo *root, RelOptInfo *baserel, Oid foreigntableid
 	if (usable_pathkeys != NIL)
 		fdwState->order_clause = orderedquery.data;
 
+	/*
+	 * Add LIMIT (FETCH N FIRST ROW ONLY) expression if not OFFSET clause
+	 * is used and that oracle version >= 12.
+	 */
```
Comment (laurenz) [inline oracle_fdw.c:956]: I think it would be fine to push down `LIMIT` even in the face of an `OFFSET`.
Then the limit passed to Oracle should be the sum of both values.
That would still give the Oracle optimizer the right clue.

File: oracle_fdw.h [hunk 5]
```diff
@@ -217,6 +217,7 @@ extern void oracleClientVersion(int *major, int *minor, int *update, int *patch,
 extern void oracleServerVersion(oracleSession *session, int *major, int *minor, int *update, int *patch, int *port_patch);
 extern void *oracleGetGeometryType(oracleSession *session);
 extern int oracleGetImportColumn(oracleSession *session, char *dblink, char *schema, char **tabname, char **colname, oraType *type, int *charlen, int *typeprec, int *typescale, int *nullable, int *key);
+extern int min_oracle_version(oracleSession *session, int major, int minor);
```
Comment (laurenz) [inline oracle_fdw.h:220]: I think that should be a static function in `oracle_fdw.c`.
It does not access the Oracle C API.

**Review by laurenz — CHANGES_REQUESTED** [review]
I have had a first quick look over your patch; here are my comments.

File: oracle_fdw.c (hunk 3 above)
Comment (darold) [reply oracle_fdw.c:778]: It is not removed it is moved to oracleGetForeignPaths() after the append of limit clause to be able to read the Oracle version information with the call to min_oracle_version() If the session information is freed before we have a problem.

File: oracle_fdw.c (hunk 1 above)
Comment (darold) [reply oracle_fdw.c:759]: There is a typo here I mean GROUP BY not ORDER BY and actually as GROUP BY are not pushed this code can be removed. Will do.

File: oracle_fdw.c (hunk 4 above)
Comment (darold) [reply oracle_fdw.c:956]: yeah, good idea.

File: oracle_fdw.h (hunk 5 above)
Comment (darold) [reply oracle_fdw.h:220]: No, you can not, there is a reference issue when calling fdwState->session->"any struct element" in oracle_fdw.c the only way I found is to declare the function in oracle_util.c, Honestly I have not taken the time to look forward at this problem.

File: oracle_fdw.c (hunk 3 above)
Comment (laurenz) [reply oracle_fdw.c:778]: Ah, ok, that makes sense. I did not see that.

File: oracle_fdw.h (hunk 5 above)
Comment (laurenz) [reply oracle_fdw.h:220]: Ok, I'll look harder later when I have more time.

File: oracle_fdw.c (hunk 1 above)
Comment (darold) [reply oracle_fdw.c:759]: Hum, no I remember now. Actually I have picked-up the code from a DB2 FDW I'm developing but I have not taking care of the current orafce_architecture and Oracle only support  a FETCH FIRST clause at end of the query. It can not be placed before and ORDER BY or a GROUP BY clause, I will remove this useless code.

**Comment by darold** [comment]
A saw that you mention that aggregate functions are not pushed down and it seems right:
```
EXPLAIN (VERBOSE on, COSTS off) SELECT sum(id) FROM typetest1;
                                          QUERY PLAN
-----------------------------------------------------------------------------------------------
 Aggregate
   Output: sum(id)
   ->  Foreign Scan on public.typetest1
         Output: id
         Oracle query: SELECT /*709894645e7b46ee7c82f53e6eeb7cac*/ r1."ID" FROM "TYPETEST1" r1
         Oracle plan: SELECT STATEMENT
         Oracle plan:   INDEX FULL SCAN TYPETEST1_PKEY
(7 rows)
```
this is also a performances killer, is there any good reason (outside the same as GROUP BY clause) why it is not pushed down?

File: oracle_fdw.h (hunk 5 above)
Comment (laurenz) [reply oracle_fdw.h:220]: I remember now - I had changed `oracleServerVersion` so that you can call that and don't need to create a new function.
See my pull request to your fork.

**Comment by laurenz** [comment]
`GROUP BY` and aggregates are not pushed down because the FDW API for that is not simple.
You'd have to add a custom scan node, if I remember right.
It is definitely on the wish list, but essentially I am waiting for someone who is motivated enough.

File: oracle_fdw.c [hunk 6]
```diff
@@ -732,6 +734,50 @@ _PG_init(void)
...
+			char *offset_val;
+
+			appendStringInfoString(&limit_clause, " FETCH FIRST ");
+			limit_val = deparseExpr(
+						fdwState->session, baserel,
+						(Expr *) root->parse->limitCount,
+						fdwState->oraTable,
+						&(fdwState->params)
+					);
+			if (root->parse->limitOffset != NULL)
```
Comment (laurenz) [inline oracle_fdw.c:763]: You check that the `LIMIT` is not NULL above; I think you should do the same for `OFFSET`.

File: oracle_fdw.c [hunk 7]
```diff
@@ -732,6 +734,50 @@ _PG_init(void)
...
+
+	if (root->parse->limitCount != NULL && !root->parse->groupClause)
+	{
+		if (IsA(root->parse->limitCount, Const) && !((Const *) root->parse->limitCount)->constisnull)
+		{
+			char *limit_val;
+			char *offset_val;
+
+			appendStringInfoString(&limit_clause, " FETCH FIRST ");
+			limit_val = deparseExpr(
```
Comment (laurenz) [inline oracle_fdw.c:757]: I am not sure if a call to `deparseExpr` is necessary here.
If it is a `Const` that is not NULL, a simple call to `datumToString` should be enough.
Such a constant has to be a `bigint` anyway.

File: oracle_fdw.c [hunk 8]
```diff
@@ -732,6 +734,50 @@ _PG_init(void)
...
+							&(fdwState->params)
+						);
+				appendStringInfo(&limit_clause, "%s+%s", limit_val, offset_val);
+			}
+			else
+				appendStringInfo(&limit_clause, "%s", limit_val);
+			appendStringInfoString(&limit_clause, " ROWS ONLY");
+		}
+	}
+	return limit_clause.data;
```
Comment (laurenz) [inline oracle_fdw.c:778]: If I read that right, you return an empty string if there is no `LIMIT` to be pushed down.
It should be NULL, if I read your modifications to `createQuery`

**Review by laurenz — CHANGES_REQUESTED** [review]

**Comment by darold** [comment]
> `GROUP BY` and aggregates are not pushed down because the FDW API for that is not simple.
> You'd have to add a custom scan node, if I remember right.
> It is definitely on the wish list, but essentially I am waiting for someone who is motivated enough.

Have you ever though about a full rewrite of oracle_fdw based on a recent architecture of postgres_fdw? Looks to me that adding custom scan node take lot of time each time, this time can be used to do this huge rewriting job but will serve to further improvement. What is your though about that?

File: oracle_fdw.c (hunk 6 above)
Comment (darold) [reply oracle_fdw.c:763]: I do not understand, line 763 show a test on offset is not null. Actually there is limitCount for the LIMIT clause and limitOffset for the OFFSET clause.

File: oracle_fdw.c (hunk 7 above)
Comment (darold) [reply oracle_fdw.c:757]: It is necessary, it can be an expression like `LIMIT nitem OFFSET prev_pos;`

Comment (darold) [reply oracle_fdw.c:757]: Hum no I'm wrong the expression result in a constant. However deparseExpr is doing the right job.

File: oracle_fdw.c (hunk 8 above)
Comment (darold) [reply oracle_fdw.c]: Fixed an pushed in commit f37b6d2

File: oracle_fdw.c (hunk 6 above)
Comment (laurenz) [reply oracle_fdw.c:763]: You check if there is a `LIMIT` clause at all, but you don't check if `IsA(root->parse->limitOffset, Const)` or if it has `->constisnull`.

File: oracle_fdw.c (hunk 7 above)
Comment (laurenz) [reply oracle_fdw.c:757]: Yes, it is doing the right job. I just wondered if the job could be done cheaper, but I admit it is not very important.

File: oracle_fdw.c (hunk 6 above)
Comment (darold) [reply oracle_fdw.c]: ok got it, fixed in commit 7720b7d as well as removing call to deparseExpr.

**Comment by laurenz** [comment]
> Have you ever though about a full rewrite of oracle_fdw based on a recent architecture of postgres_fdw? Looks to me that adding custom scan node take lot of time each time, this time can be used to do this huge rewriting job but will serve to further improvement. What is your though about that?

There are a couple of things that could be done better for sure.
I lack the initiative to tackle a project of that size though.

**Comment by laurenz** [comment]
I found a problem with the `LIMIT` pushdown patch:

```
EXPLAIN (COSTS OFF) SELECT count(*) FROM typetest1 LIMIT 1;
                                                       QUERY PLAN
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
 Limit
   ->  Aggregate
         ->  Foreign Scan on typetest1
               Oracle query: SELECT /*5d31d5dda80a4849b792294e4bd31b17*/ '1' FROM "TYPETEST1" r1 FETCH FIRST 1 ROWS ONLY
(4 rows)
```

We cannot push down `LIMIT` if there is an aggregate function involved, even if there is no `GROUP BY`.

**Comment by darold** [comment]
Right, I remember now what was the utility to scan the target list in my first patch except that this was with aggregate push down. I also fixed a bug in my previous patch.

File: oracle_fdw.c [hunk 9]
```diff
@@ -752,19 +752,37 @@ deparseLimit(PlannerInfo *root, struct OracleFdwState *fdwState, RelOptInfo *bas
...
 			{
-				offset_val = datumToString(((Const *) root->parse->limitOffset)->constvalue, ((Const *) root->parse->limitOffset)->consttype);
-				appendStringInfo(&limit_clause, "%s+%s", limit_val, offset_val);
+				Node *ptl = (Node *) lfirst(lc);
+				if (IsA(ptl, TargetEntry))
+				{
+					TargetEntry *tge = (TargetEntry *) ptl;
+					if (IsA(tge->expr, Aggref))
+						add_limit_clause = false;
+				}
```
Comment (laurenz) [inline oracle_fdw.c:767]: That is not good enough; it only checks if the top-level node is an aggregate. You have to check if there is an aggregate anywhere in the plan:

```
EXPLAIN (COSTS OFF) SELECT 12 - count(*) FROM typetest1 LIMIT 1;

 Limit
   ->  Aggregate
         ->  Foreign Scan on typetest1
               Oracle query: SELECT /*5d31d5dda80a4849b792294e4bd31b17*/ '1' FROM "TYPETEST1" r1 FETCH FIRST 1 ROWS ONLY
(4 rows)
```

**Review by laurenz — CHANGES_REQUESTED** [review]

**Comment by laurenz** [comment]
There is yet another problem: we get wrong results if `WHERE` conditions are not pushed down:

```
EXPLAIN (COSTS OFF) SELECT id FROM typetest1 WHERE vc < 'u' LIMIT 1;

 Limit
   ->  Foreign Scan on typetest1
         Filter: ((vc)::text < 'u'::text)
         Oracle query: SELECT /*da1e34a7adb79160f2290e7b654c1d43*/ r1."ID", r1."VC" FROM "TYPETEST1" r1 FETCH FIRST 1 ROWS ONLY
(4 rows)
```

Fortunately that should be easy to fix:
You can check if there are any `fdwState->local_conds` and not push down in that case.

File: oracle_fdw.c (hunk 9 above)
Comment (darold) [reply oracle_fdw.c:767]: > That is not good enough; it only checks if the top-level node is an aggregate. You have to check if there is an aggregate anywhere in the plan:
> 
> ```
> EXPLAIN (COSTS OFF) SELECT 12 - count(*) FROM typetest1 LIMIT 1;
> 
>  Limit
>    ->  Aggregate
>          ->  Foreign Scan on typetest1
>                Oracle query: SELECT /*5d31d5dda80a4849b792294e4bd31b17*/ '1' FROM "TYPETEST1" r1 FETCH FIRST 1 ROWS ONLY
> (4 rows)
> ```

The main problem here is that aggregates must be pushed down, adding additional work to deparse OpExpr to handle this case adds development time that could be put into adding aggregate push down :-( As you see with this "ectoplasmic" patch I lack time to get it seriously and actually I was not aware that oracle_fdw doesn't push down GROUP BY and aggregates otherwise I would not have started this patch. I will have more time probably in June I will get back to this PR.

Comment (laurenz) [reply oracle_fdw.c:767]: I see. I guess that even if aggregates were pushed down, not all aggregates could be (`array_agg`), so this would still remain an issue.
I understand that you have limited capacity to work on this; so have I.
Thanks a lot for your contribution so far. If you don't beat me to it, I will try to pick up the ball where you left it.
It would be a pity to let this go to waste.

**Comment by laurenz** [comment]
Detecting aggregate functions was easier than I thought.
